managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1850,9 +1854,172 @@ public void offloadFailed(ManagedLedgerException e, Object ctx) {
+                                    } else {
+                                        callback.offloadComplete(firstUnoffloaded, ctx);
+                                    }
+                                });
+                    } else {
+                        // All failed, so throw up the error
+                        callback.offloadFailed(new ManagedLedgerException(offloadException), ctx);
+                    }
+                });
     }
 
+    private CompletableFuture<Void> updateLedgerInfoForOffloaded(List<Pair<Long, byte[]>> contexts) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        synchronized (this) {
+            ledgersListMutex.lock();
+
+            contexts.forEach((context) -> {
+                    long ledgerId = context.getLeft();
+                    LedgerInfo oldInfo = ledgers.get(ledgerId);
+                    if (oldInfo == null) {
+                        cleanupOffloadedOnFailure(context.getLeft(), context.getRight(),
+                                                  "Trimmed while offload in progress");
+                    } else {
+                        LedgerInfo newInfo = oldInfo.toBuilder()
+                            .setOffloadContext(ByteString.copyFrom(context.getRight())).build();
+                        ledgers.put(ledgerId, newInfo);
+                    }
+                });
+
+            log.info("[{}] Updating of ledgers list after offloading", name);
+            store.asyncUpdateLedgerIds(name, getManagedLedgerInfo(), ledgersStat,
+                    new MetaStoreCallback<Void>() {
+                        @Override
+                        public void operationComplete(Void result, Stat stat) {
+                            log.info("[{}] End Offload. ledgers={} totalSize={}", name, ledgers.size(),
+                                     TOTAL_SIZE_UPDATER.get(ManagedLedgerImpl.this));
+                            ledgersStat = stat;
+                            ledgersListMutex.unlock();
+
+                            promise.complete(null);
+
+                            contexts.forEach((context) -> {
+                                    long ledgerId = context.getLeft();
+                                    ledgerCache.remove(ledgerId);
+                                    log.info("[{}] Removing offloaded ledger from bookkeeper {}",
+                                             name, ledgerId);
+                                    asyncDeleteLedger(ledgerId);
+                                });
+                        }
+
+                        @Override
+                        public void operationFailed(MetaStoreException e) {
+                            log.warn("[{}] Failed to update the list of ledgers after offloading", name, e);
+                            ledgersListMutex.unlock();
+                            contexts.forEach(context -> cleanupOffloadedOnFailure(
```

Comment:
hmm.. this can be potentially very dangerous. updating metadata is "cheaper" than offloading data (in terms of resource usage), if we successfully offload data but hit a hiccup on updating metadata, this call here basically trashes all the successfully offloaded ledgers. I think `asyncUpdateLedgerIds` should be put in a `back-off` retry logic before claiming it failed.